[지구] 리뷰용 PR

Ojeegu opened the pull request:
- step2 구현 중
- 회원 정보 수정은 아직 구현중

Ojeegu committed:
Init : thymeleaf 의존성 주입 및 템플릿 엔진 설정 추기

build.gradle 파일에 thymeleaf 의존성 주입 후 application.yml 파일에 템플릿 엔진 설정 추가. 가독성을 위해 .properties 대신 .yml을 사용.

Ojeegu committed:
feat : 웹페이지 HomeController 생성

localhost:8081/ 이동 시 홈페이지 첫 화면(index.html)이 보이도록 controller 생성

Ojeegu committed:
feat : 사용자 정보를 저장하고 관리하는 User 객체 생성

getter, setter 메서드 추가

Ojeegu committed:
feat : 회원가입 시 사용자 정보를 map에 저장하는 기능 및 email로 사용자 검색하는 기능 추가

사용자가 회원가입 하면 자동으로 고유 id값을 부여

Ojeegu committed:
feat : 회원가입 및 회원 목록을 볼 수 있는 기능 추가

회원가입 시 자동으로 userMap에 회원 정보가 추가되고, 이를 토대로 회원가입한 회원들의 목록을 확인 가능.

Ojeegu committed:
style : 홈페이지 html 및 css 파일 변경

기존 파일이 오류가 나서 다른 템플릿을 다운 받아 업데이트

Ojeegu committed:
style : 공통 부분 layout/nav.html로 따로 빼내어 한번에 관리하도록 수정

공통으로 들어가는 navigation 부분을 layout 폴더에 nav.html 파일로 따로 빼내서 추가 및 수정 시 한번에 관리할 수 있도록 만듦.

Ojeegu committed:
refactor : UserDTO 객체 생성 및 DTO 객체를 통해 data 전송하도록 수정

원본 데이터 보안 유지를 위해 User 객체를 그대로 사용하지 않고 UserJoinDTO와 UserReadDTO를 사용해 데이터를 전송.

Ojeegu committed:
feat : 회원목록 페이지에서 사용자 닉네임 클릭 시 회원정보 페이지로 이동하도록 기능 추가

Thymeleaf 문 이용하여 클릭한 회원 id에 해당하는 페이지로 이동하도록 기능 추가

Ojeegu committed:
style : 회원 정보 페이지 추가

회원 정보를 보여주는 페이지

Ojeegu committed:
style : css 파일 추가

페이지 템플릿 적용

Ojeegu committed:
feat : 회원 정보 페이지로 이동하는 기능 추가

 회원의 id를 주소로 받아  회원 정보 페이지로 이동

Ojeegu committed:
docs : Update readMe file

step1 완료 후 readMe 파일 갱신

Ojeegu committed:
refactor : 불필요한 필드 제거 및 동시성 문제 해결

UserJoinDTO의 id 필드 삭제, toUser 메서드 setId 부분 삭제, 동시성 문제를 해결하기 위해 HashMap과 Long id 대신 ConcurrentHashMap, AtomicLong 사용.

Ojeegu committed:
feat : 글쓰기 기능 추가

글 작성 시 글을 등록하고 메인 화면으로 이동하는 기능을 추가. 메인 화면에서는 글 목록을 확인할 수 있다.

Ojeegu committed:
feat : 상세 글 보기 기능 추가

글 목록에서 글 제목 클릭 시 상세 보기 페이지로 이동.

Ojeegu committed:
feat : 상세 글 보기 기능 추가(2)

show.html 파일 누락분

Ojeegu committed:
style : 기본 페이지 디자인 변경

글목록 상단에 작성자 항목 추가

Ojeegu committed:
style : 글 상세 보기 URL 경로 변경

글 상세 보기 URL을 좀 더 명확하게 수정

Ojeegu committed:
docs : Update readME file

step2 구현 중 readME 파일 업데이트

Ojeegu committed:
feat : 게시글 등록 시 내림차순 정렬 기능 추가

최신 순서대로 보일 수 있도록 내림차순 정렬 기능 추가

Ojeegu committed:
style : 상세 글 댓글 예시 수정

상세 글 보기에서 댓글 예시 내용을 수정

Ojeegu committed:
refactor : 매개변수 앞에 final 키워드 추가

매개변수 앞에 final 키워드를 추가하여 외부에서 수정하는 실수/오류를 방지, 코드의 안정성을 높임.

tmdgusya committed:
Merge pull request #43 from Ojeegu/step1

[지구] Step-1,2 구현

Ojeegu committed:
feat : 데이터베이스 의존성 설정

H2 데이터베이스 의존성 및 기본 설정

Ojeegu committed:
feat : 데이터베이스 저장할 테이블 생성

USER 테이블 및 각각의 정보를 저장할 칼럼 생성

Ojeegu committed:
feat : database repository로 번경

Jdbc template을 이용한 repository 생성 후 기존의 memeoryUserRepository 삭제

Ojeegu committed:
style : 변수명 변경

JdbcTemplateUserRepository 변수명 직관적으로 변경

Ojeegu committed:
feat : Article 및 ArticleDTO 필드에 작성자 추가

작성자 정보를 나타내는 기능을 구현하기 위해 필드에 작성자(author) 추가

Ojeegu committed:
feat : 게시글 데이터베이스 테이블 생성

게시글 데이터 저장을 위해 게시글용 테이블 생성

Ojeegu committed:
feat : 게시글 update 기능 추가

게시글을 수정할 수 있는 update 기능을 추가(아직 구현은x)

Ojeegu committed:
feat : 게시글 데이터베이스에 저장 및 수정하는 기능 추가

객체를 이용하여 SQL문을 통해 데이터베이스 테이블에 저장하고 수정하는 기능을 추가

Ojeegu committed:
remove : 기존 게시글 repository 삭제

필요 없는 repository 삭제

Ojeegu committed:
feat : 게시글 수정 기능 구현

게시글 원래 정보를 불러와 수정하고 덮어쓰우는 기능을 구현

Ojeegu committed:
feat : 게시글 수정 페이지 추가 및 URL 연결

게시글 수정 페이지를 추가하고 URL을 매핑(get, post). 수정 완료 후에는 게시글 상세페이지로 이동하는 것으로 구현

Ojeegu committed:
feat : 상세보기 페이지에서 수정 및 글목록으로 이동하는 기능 추가

상세보기 페이지에 수정, 삭제, 글목록 버튼 추가하고 css 디자인 수정

devbattery commented on src/main/java/kr/codesqaud/cafe/controller/ArticleController.java:
```diff
@@ -0,0 +1,53 @@
+    }
+
+    @PostMapping("/post/write")
+    public String writePost(@ModelAttribute final ArticleDTO articleDto) {
```
매핑하실 때 매개변수를 전부 final로 하신 거 같은데 그 이유는 무엇인가요??

devbattery replied:
> 보안상의 이유. 컨트롤러의 모든 매개변수 final로 바꾸는 게 좋을까? 해서 했음

devbattery commented on src/main/java/kr/codesqaud/cafe/controller/dto/ArticleDTO.java:
```diff
@@ -0,0 +1,109 @@
+        this.author = author;
+    }
+
+    public static ArticleDTO toArticleDTO(final Article article) {
+        return new ArticleDTO(article.getId(), article.getTitle(), article.getContent(), article.getAuthor());
+    }
```
저는 생성자에서 바로 넣어줬는데
다른 분들의 코드도 그렇게 이런 식으로 리턴하는 메서드 작성하셨더라구요
혹시 생성자에서 바로 넣어주는 것과 이렇게 넣어주는 것의 차이가 궁금합니다..!

devbattery replied:
> 생성자로 넣어줄 필요가 없는데도 final로 선언하기 위해 생성해주는 게 맞나?

devbattery commented on src/main/java/kr/codesqaud/cafe/repository/JdbcTemplateArticleRepository.java:
```diff
@@ -0,0 +1,67 @@
+        return wantedPost.stream().findAny();
+    }
+
+    public RowMapper<Article> articleRowMapper() {
+        return (rs, rowNum) -> {
+            Article article = new Article();
+            article.setId(rs.getLong("ID"));
+            article.setTitle(rs.getString("TITLE"));
+            article.setContent(rs.getString("CONTENT"));
+            article.setAuthor(rs.getString("AUTHOR"));
+            return article;
+        };
+    }
```
UserRepository도 그렇고 RowMapper() 메서드는
`BeanPropertyRowMapper.newInstance()`를 쓰면 한줄로 처리 가능해요!
이 코드 같은 경우는

```java
    public RowMapper<Article> articleRowMapper() {
        return BeanPropertyRowMapper.newInstance(Article.class);
    }
```

이렇게 짜시면 한줄로 해결됩니다!

devbattery commented on src/main/java/kr/codesqaud/cafe/repository/JdbcTemplateArticleRepository.java:
```diff
@@ -0,0 +1,67 @@
+        jdbcTemplate = new JdbcTemplate(dataSource);
+    }
+
+    @Override
+    public void save(Article article) {
+        SimpleJdbcInsert jdbcInsert = new SimpleJdbcInsert(jdbcTemplate);
+        jdbcInsert.withTableName("ARTICLE_TB").usingGeneratedKeyColumns("ID");
+
+        Map<String, Object> articleParameters = new HashMap<>();
+        articleParameters.put("TITLE", article.getTitle());
+        articleParameters.put("CONTENT", article.getContent());
+        articleParameters.put("AUTHOR", article.getAuthor());
+
+        Number key = jdbcInsert.executeAndReturnKey(new MapSqlParameterSource(articleParameters));
+        article.setId(key.longValue());
+    }
```
SimpleJdbcInsert 사용하실 거라면 필드로 생성하셔서 생성자로 주입 받은 다음에

```java
SqlParameterSource param = new BeanPropertySqlParameterSource(article);
Number key = simpleJdbcInsert.executeAndReturnKey(param);
user.setId(key.longValue());
```

이런 식으로 하면 3줄로 끝나요!

devbattery commented on src/main/java/kr/codesqaud/cafe/service/ArticleService.java:
```diff
@@ -0,0 +1,45 @@
+        this.articleRepository = articleRepository;
+    }
+
+    public void write(final ArticleDTO articleDto) {
```
여기 서비스에서도 매개변수로 DTO를 받으셨네요
제 로직에서는 Article 도메인이 넘어오는데
지구님에게 dto 사용 법 좀 제대로 배워야겠네요..

devbattery commented on src/main/java/kr/codesqaud/cafe/service/ArticleService.java:
```diff
@@ -0,0 +1,45 @@
+        articleRepository.update(originArticle);
+    }
+
+    public List<ArticleDTO> gatherPosts() {
+        List<Article> postList = articleRepository.gatherAll();
+        List<ArticleDTO> postDTOList = new ArrayList<>();
+        for(Article article : postList) {
+            postDTOList.add(ArticleDTO.toArticleDTO(article));
+        }
+        return postDTOList;
+    }
```
저도 dto로 필터링 해줄 때 이렇게 해줬는데
List 2개 만들고 for문 돌리는 게 최선일까요?

devbattery commented on src/main/java/kr/codesqaud/cafe/service/UserService.java:
```diff
@@ -0,0 +1,56 @@
+        this.userRepository = userRepository;
+    }
+
+    public void signUp(final UserJoinDTO userJoinDTO) {
+        //1. UserJoinDTO -> User 변환
+        User user = User.toUser(userJoinDTO);
+
+        //2. repository signUp 메서드 호출
+        userRepository.join(user);
+    }
```
아 계속 DTO로 쓰다가 여기에서 User로 변환해주는 거군요..!

devbattery commented on src/main/resources/schema.sql:
```diff
@@ -0,0 +1,17 @@
+    EMAIL    VARCHAR(100) NOT NULL
+);
+
+CREATE TABLE IF NOT EXISTS ARTICLE_TB
+(
+    ID         BIGINT PRIMARY KEY AUTO_INCREMENT,
+    TITLE      VARCHAR(100) NOT NULL,
+    CONTENT    LONGTEXT     NOT NULL,
+    AUTHOR     VARCHAR(50)  NOT NULL
+);
```
이제 여기에 TIMESTAMP도 추가하시면 더 완벽할 거 같아요!

devbattery reviewed (commented):
코드 읽으면서 많이 깨달았습니다..!
dto는 이렇게 쓰는 거구나를 깨달았어요

Ojeegu committed:
refactor : ArticleRepository NamedParameterJdbcTemplate로 변경

기존 JdbcTemplate에서 NamedParameterJdbcTemplate로 변경. 이 방법이 유지보수에 용이하고(순서 상관x) 요즘 보편적으로 사용된다고 한다.

Ojeegu committed:
refactor : UserRepository NamedParameterJdbcTemplate로 변경

기존에 JdbcTemplate 쓰던 것을 NamedParameterJdbcTemplate로 변경

Ojeegu committed:
refactor : @ModelAttribute 어노테이션 삭제

생략해도 되는 어노테이션을 삭제하고 html 파일을 수정

Ojeegu committed:
feat : 회원정보 수정 기능 추가

회원 상세 페이지로 들어가 수정하기 버튼을 누르면 이름과 이메일을 수정할 수 있도록 기능 추가. 수정완료 버튼을 누르면 갱신된 회원 정보 상세 페이지로 이동한다.

Ojeegu committed:
docs : Update readME file

readME 파일 업데이트
